### PR TITLE
feat: Add payload validation [SIW-284]

### PR DESCRIPTION
**Pull request by grausof**
## Short description
Added function for validating the JWT containing:
- JWS signature validation
- JWT payload decoding
- JWT claim set

Also added `typedoc` for generating documentation.

## How to test
Added unit test

### CODE REVIEW

File: src/__tests__/jwt_claims_set.test.ts [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+const jwtPayload: JWTPayload = {
+  iss: 'demo',
+  sub: 'demo',
+  iat: 1689236635,
+  exp: 1689240235,
+};
+
+describe('Basic JWT Claims Set verification', function () {
+  it('must check the parameters correctly', async () => {
+    let verified = verifyJwtClaimSet(jwtHeader, jwtPayload, {
```
Comment (hevelius) [inline src/__tests__/jwt_claims_set.test.ts]:
```suggestion
    const verified = verifyJwtClaimSet(jwtHeader, jwtPayload, {
```

File: src/__tests__/jwt_claims_set.test.ts [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+    });
+    expect(verified).toBe(jwtPayload);
+  });
+
+  it('must fail the check because the JWT has expired', async () => {
+    try {
+      verifyJwtClaimSet(jwtHeader, jwtPayload);
+    } catch (err) {
+      expect(err).toBeInstanceOf(JWTExpired);
+    }
```
Comment (hevelius) [inline src/__tests__/jwt_claims_set.test.ts]:
```suggestion
   expect(() => verifyJwtClaimSet(jwtHeader, jwtPayload)).toThrowError(
      JWTExpired
    );
```

File: src/__tests__/jwt_claims_set.test.ts [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+  });
+
+  it('must fail the check because the typ is wrong', async () => {
+    try {
+      verifyJwtClaimSet(jwtHeader, jwtPayload, {
+        typ: 'invalid',
+      });
+    } catch (err) {
+      expect(err).toBeInstanceOf(JWTClaimValidationFailed);
+    }
```
Comment (hevelius) [inline src/__tests__/jwt_claims_set.test.ts]:
```suggestion
    expect(() =>
      verifyJwtClaimSet(jwtHeader, jwtPayload, {
        typ: 'invalid',
      })
    ).toThrowError(JWTClaimValidationFailed);
```

File: src/index.tsx [hunk 4]
```diff
@@ -18,10 +24,93 @@ const IoReactNativeJwt = NativeModules.IoReactNativeJwt
...
+ *
+ * const { payload, protectedHeader } = await decode(jwt)
+ *
+ * console.log(protectedHeader)
+ * console.log(payload)
```
Comment (hevelius) [inline src/index.tsx]:
```suggestion
```

File: src/index.tsx [hunk 5]
```diff
@@ -18,10 +24,93 @@ const IoReactNativeJwt = NativeModules.IoReactNativeJwt
...
+ *
+ * const isSignatureValid = await verify(jwt, jwk)
+ *
+ * console.log(isSignatureValid)
```
Comment (hevelius) [inline src/index.tsx]:
```suggestion
```

File: src/index.tsx [hunk 6]
```diff
@@ -18,10 +24,93 @@ const IoReactNativeJwt = NativeModules.IoReactNativeJwt
...
+export const verify = async (
+  token: string,
+  jwk: JWK,
+  options: JWTClaimVerificationOptions = {}
+): Promise<JWTDecodeResult> => {
+  const { protectedHeader, payload } = await decode(token);
+  const signatureIsValid = await isSignatureValid(token, jwk);
+
+  if (signatureIsValid) {
+    let verifiedPayload = jwtPayload(protectedHeader, payload, options);
```
Comment (hevelius) [inline src/index.tsx]:
```suggestion
    const verifiedPayload = jwtPayload(protectedHeader, payload, options);
```

File: src/index.tsx [hunk 7]
```diff
@@ -18,10 +24,93 @@ const IoReactNativeJwt = NativeModules.IoReactNativeJwt
...
+ *
+ * const { payload, protectedHeader } = await verify(jwt, jwk, {
+ *   typ: 'entity-statement+jwt',
+ *   requiredClaims: ['iss', 'sub', 'metadata'],
+ * })
+ *
+ * console.log(protectedHeader)
+ * console.log(payload)
```
Comment (hevelius) [inline src/index.tsx]:
```suggestion
```

File: src/jwt_claims_set.ts [hunk 8]
```diff
@@ -0,0 +1,196 @@
...
+
+const checkAudiencePresence = (audPayload: unknown, audOption: unknown[]) => {
+  if (typeof audPayload === 'string') {
+    return audOption.includes(audPayload);
+  }
+
+  if (Array.isArray(audPayload)) {
+    /*
+     * Each principal intended to process the JWT MUST
+     *identify itself with a value in the audience claim
```
Comment (hevelius) [inline src/jwt_claims_set.ts]:
```suggestion
     * identify itself with a value in the audience claim
```

File: src/utils/errors.ts [hunk 9]
```diff
@@ -0,0 +1,196 @@
...
+    return 'ERR_JOSE_GENERIC';
+  }
+
+  /** A unique error code for the particular error subclass. */
+  code: string = 'ERR_JOSE_GENERIC';
+
+  constructor(message?: string) {
+    super(message);
+    this.name = this.constructor.name;
+    // @ts-ignore
```
Comment (hevelius) [inline src/utils/errors.ts:24]: Is this required?

File: src/utils/secs.ts [hunk 10]
```diff
@@ -0,0 +1,51 @@
+const minute = 60;
+const hour = minute * 60;
+const day = hour * 24;
+const week = day * 7;
+const year = day * 365.25;
+
+const REGEX =
```
Comment (hevelius) [inline src/utils/secs.ts:12]: What do you think to add a little comment to explain what this regex does?

**Review by hevelius — CHANGES_REQUESTED** [review]

File: src/utils/errors.ts (hunk 9 above)
Comment (grausof) [reply src/utils/errors.ts:24]: It's useful if you want to throw a generic error. For now we don't have this case but I preferred to insert it anyway to make it the same as the others!

File: src/utils/secs.ts (hunk 10 above)
Comment (grausof) [reply src/utils/secs.ts:12]: of course, thanks!

**Review by hevelius — APPROVED** [review]
LGTM

**Pull request merged**